src/log.c

```diff
@@ -43,7 +56,7 @@ char* log_message_timestamp(char* dest, size_t maxlen) {
     return dest;
 }
 
-void log_message_internal(const char* tag, log_level_t level, const char* message, va_list args, FILE* out) {
+void log_message_internal_printer(const char* tag, log_level_t level, const char* message, va_list args, FILE* out) {
```

Comment:
Every function call slows execution. It should be inline.

Reply:
When built for release the build system used -O3 and as part of that optimisation set the inlining is automatically performed if the function is considered worth doing so.

If you check the ASM produced by GCC using godbolt
https://www.godbolt.org/z/iNQAxZ

I copied and pasted there the vast majority of log.c/log.h and added a very simple main to invoke log_message.

You will be able to see that for example the compiler is already performing the inlining of, basically, all the functions:
- it log_message_timestamp in log_message_internal_printer
- inlined the "new" version of log_message_internal_printer in log_message_internal
- inlined the "new" version of "log_message_internal" in "log_message.

In godbolt there isn't any code call at all in the body of log_message because of the automatic inlining.

I have slightly changed the code in godbolt to avoid that GCC drops entire branches of the code because there is nothing changing log_sinks_registered_count and in addition I dropped the body of log_level_to_string because it's not really interesting for the discussion, makes only a lot of noise in the final asm.

But the compiler balance the automatic inlining with the size of the function, if it's a very often called function and it's way too big (it's not the case) the CPU may not be able to keep it in the L1I cache (the L1 instruction set cache) and it would have to swap it out continuously during an execution unit basically destroying any resemble of performance.

Even when you mark a function as to be "inline" that is just an hint to the compiler, it will not do so if it will consider it not the best option. To enforce the inlining of functions there are apposite attributes like "always_inline" specific to GCC.

The only real difference between marking the functions as inline or not is that the final binary may be reduced because instead of keeping the various copies of the functions in the object file and in the final executable they are dropped because the compiler expects that they are not in use anywhere else, but this doesn't affect the execution speed and the size of the binary itself is affected really marginally (unless the very same approach is applied to every function making the code hard to refactor if needed).

Reply:
Not only. When calling function compiler have to take a snapshot of rbp register so it executes a few additional instructions. So it makes little difference on performance.

Reply:
I was referring to the usage of the inline attribute, if GCC decides to inline and does so when using the inline attributes it drops the original function in the resulting object file.

Yes, of course a function call implies a number of registry settings a jumps in the memory that have a direct hit on the perfs.

But the inlining is already performed automatically so it, I think, be ok to leave it as it is.